src/leveling.jl

```diff
@@ -0,0 +1,220 @@
+import Base.isless
+import Base.isgreater
+import DataStructures.PriorityQueue, DataStructures.enqueue!, DataStructures.dequeue_pair!
+
+"""
+    low_leveling(op, marker, mask; [dims])
+    low_leveling(op, marker, mask, se)
+
+A function g is an lower leveling of a function f if and only if 
+for any couple of neighbouring pixels (p, q) g(p)>g(q)->g(p)>=f
+This algorithm modify marker to become a high_leveling of ref
+
+The `dims` keyword is used to specify the dimension to process by constructing the box shape
+structuring element [`strel_box(marker; dims)`](@ref strel_box). For generic structuring
+element, the half-size is expected to be either `0` or `1` along each dimension.
+
+# See also
+ [`high_leveling!`](@ref),[`leveling`](@ref),
+ [`underbuild`](@ref),[`overbuild`](@ref)
+ 
+# References
+- [1] 'From connected operators to levelings' [Meyer,1998] In ISMM '98: Proceedings of the fourth international symposium on Mathematical morphology and its applications to image and signal processingJune 1998 Pages 191–198
+- [2] 'Mathematical Morphology: From Theory to Applications' chap6 [Serra et al.] https://doi.org/10.1002/9781118600788.ch8 
+"""
+function low_leveling(ref, marker; dims=coords_spatial(marker))
+    return low_leveling(ref, marker, strel_box(marker, dims))
+end
+function low_leveling(ref, marker, se)
+    return low_leveling!(similar(ref), ref, marker, se)
+end
+
+function low_leveling!(out, ref, marker; dims=coords_spatial(marker))
+    return low_leveling!(out, ref, marker, strel_box(marker, dims))
+end
+function low_leveling!(out, ref, marker, se)
+    return _low_leveling!(out, ref, marker, se)
+end
+function _low_leveling!(out::AbstractArray{T,N}, ref::AbstractArray{T,N}, marker::AbstractArray{T,N}, se) where {T<:ImageCore.NumberLike,N}
+    axes(out) == axes(ref) == axes(marker) || throw(DimensionMismatch("images should have the same axes"))
+
+    # For generic structuring element, the half-size should to be either `0` or `1` along
+    # each dimension. See also section 6.2.3 of [2].
+    se_size = strel_size(se)
+    if length(se_size) != N
+        msg = "the input structuring element is not for $N dimensional array, instead it is for $(length(se_size)) dimensional array"
+        throw(DimensionMismatch(msg))
+    end
+    se = strel(CartesianIndex, se)
+    out = copy(marker)
+
+    # declare priority queue, store position of pixel associated with priority 
+    pq = PriorityQueue{CartesianIndex{N},T}()
+    R = CartesianIndices(axes(out))
+    for i in R
+        @inbounds if out[i] < ref[i]
+            @inbounds current_max = out[i]
+            for Δi in se # examine neighborhoods
+                ii = i + Δi
+                if checkbounds(Bool, R, ii) #check that we are in the image
+                    @inbounds current_max = max(current_max, out[ii])
```

Comment:
IIUC, when you add an `@inbounds` at the outer place, then it applies to all the inner regions. Thus you don't need to add `@inbounds` for every place.

This holds true as long as it is the same function. That said, if there's another function that you want to pass `@inbounds`, you can use `Base.@propagate_inbounds` to define that _inline_-ed function:

```julia
@inline f(x, i) = x[i]
Base.@propagate_inbounds g(x, i) = x[i]

hf() = @inbounds f([1], 2) # this errors
hf()

hg() = @inbounds g([1], 2) # this doesn't
hg()
```

The `@inbounds` macro only applies when it's inside a function, which is why we need `hg()`/`hf()` in above example.